### PR TITLE
Problem: error return of closing writable file is not checked

### PR DESCRIPTION
**Pull request by mmsqe**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Closes: https://github.com/crypto-org-chain/cronos/issues/854
# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Comment by yihuang** [comment]
Reading is fine, we only need to handle writing cases

File: versiondb/client/changeset.go [hunk 1]
```diff
@@ -334,13 +334,16 @@ func SortFilesByFirstVerson(files []string) ([]FileWithVersion, error) {
...
-func ReadFirstVersion(fileName string) (uint64, error) {
+func ReadFirstVersion(fileName string) (value uint64, returnErr error) {
 	fp, err := openChangeSetFile(fileName)
 	if err != nil {
 		return 0, err
 	}
-	defer fp.Close()
-
+	defer func() {
+		if err := fp.Close(); returnErr == nil {
```
Comment (mmsqe) [inline versiondb/client/changeset.go:343]: > Reading is fine, we only need to handle writing cases

yea, I was about to ask if need for read

File: versiondb/client/dump.go [hunk 2]
```diff
@@ -201,7 +201,10 @@
 		return err
 	}
 
-	return writer.Flush()
+	if err := writer.Flush(); err != nil {
+		return err
+	}
+	return fp.Sync()
```
Comment (yihuang) [inline versiondb/client/dump.go:207]: But call sync on these temporary files hurt performance for no good though.

**Review by yihuang — COMMENTED** [review]
Ideally, we don't sync the intermediate files, it hurt performance for no good

File: versiondb/client/dump.go (hunk 2 above)
Comment (mmsqe) [reply versiondb/client/dump.go]: so we can only detect invalid file when read

Comment (yihuang) [reply versiondb/client/dump.go]: if the process don't finish succesfully, user just need to rerun.

File: versiondb/client/dump.go [hunk 3]
```diff
@@ -191,7 +191,12 @@ func dumpRangeBlocks(outputFile string, tree *iavl.ImmutableTree, blockRange Ran
 	if err != nil {
 		return err
 	}
-	defer fp.Close()
+	defer func() {
+		cerr := fp.Close()
+		if err == nil {
+			err = cerr
```
Comment (yihuang) [inline versiondb/client/dump.go:197]: I think we need named return variable, to modify the return err here.

**Review by yihuang — APPROVED** [review]

**Comment by yihuang** [comment]
I want to wait for #845 to merge first, that test takes long time to rerun.

**Pull request merged**